File: csharp/Structural/09-Decorator/Decorator/UrgentDecorator.cs
using Decorator.Component;

namespace Decorator.Decorator;

public class UrgentDecorator : JobCardDecorator
{
    private const decimal SurchargeRate = 0.25m;

    public UrgentDecorator(IJobCard jobCard) : base(jobCard) { }

    public override decimal Cost => _wrapped.Cost * (1 + SurchargeRate);
    public override string Description => $"🚨 URGENT | {_wrapped.Description}";

    public override void Execute()
    {
        Console.WriteLine("  📟 Paging on-call handyman immediately...");
        base.Execute();
        Console.WriteLine("  ⚡ Urgent flag set — supervisor notified.");
    }
}

File: typescript/Structural/09-Decorator/Decorator/UrgentDecorator.ts
import { IJobCard } from "../Component/IJobCard";
import { JobCardDecorator } from "./JobCardDecorator";

export class UrgentDecorator extends JobCardDecorator {
  private static readonly SURCHARGE_RATE = 0.25;

  constructor(jobCard: IJobCard) {
    super(jobCard);
  }

  get cost(): number {
    return this.wrapped.cost * (1 + UrgentDecorator.SURCHARGE_RATE);
  }

  get description(): string {
    return `🚨 URGENT | ${this.wrapped.description}`;
  }

  execute(): void {
    console.log("  📟 Paging on-call handyman immediately...");
    super.execute();
    console.log("  ⚡ Urgent flag set — supervisor notified.");
  }
}
